src/IssueViz.Security.UnitTests/ListenerFactoryTests.cs

```diff
@@ -0,0 +1,146 @@
+        public void Create_FirstAvailablePortIsUsed(int startPort, int endPort, int firstFreePort)
+        {
+            IListenerFactory testSubject = new ListenerFactory(new TestLogger(logToConsole: true));
+
+            bool isFreePortInRange = firstFreePort >= startPort && firstFreePort <= endPort;
+
+            var allPorts = CreateRange(startPort, endPort);
+
+            // If any of the ports required by the test are not available then bail out.
+            if(!AreAllPortsAvailable(allPorts))
+            {
+                Assert.Inconclusive("Test setup error: some test ports are in use.");
+            }
+
+            // The unavailable ports are all at the start of the range.
+            var unavailablePorts = CreateRange(startPort, firstFreePort - 1);
+            var remainingFreePorts = CreateRange(firstFreePort + 1, endPort);
+
+            using(new PortGrabber(unavailablePorts))
+            using (var actual = testSubject.Create(startPort, endPort))
+            {
+                if (isFreePortInRange)
+                {
+                    var expectedPort = GetPrefix(firstFreePort);
+                    actual.Should().NotBeNull();
+                    actual.IsListening.Should().BeTrue();
+                    actual.Prefixes.Count.Should().Be(1);
+                    actual.Prefixes.First().Should().Be(expectedPort);
+                    IsPortAvailable(firstFreePort).Should().BeFalse();
+                }
+                else
+                {
+                    actual.Should().BeNull();
+                }
+
+                AreAllPortsAvailable(remainingFreePorts).Should().BeTrue();
+            }
+
+            // The port used by the test subject should have been released on disposal
+            IsPortAvailable(firstFreePort).Should().BeTrue();
+        }
+
+        private static int[] CreateRange (int start, int end)
+        {
+            if (start > end)
+            {
+                return Array.Empty<int>();
+            }
+            return Enumerable.Range(start, end - start + 1).ToArray();
+        }
+
+        private static bool AreAllPortsAvailable(params int[] ports) =>
+            ports.All(IsPortAvailable);
+
+        private static bool IsPortAvailable(int port)
+        {
+            try
+            {
+                using (var listener = new HttpListener())
+                {
+                    listener.Prefixes.Add(GetPrefix(port));
+                    listener.Start();
+                }
+            }
+            catch (Exception)
+            {
+                return false;
+            }
+            return true;
+        }
+
+        private static string GetPrefix(int port) => $"http://localhost:{port}/";
+
+        /// <summary>
+        /// Helper class to listen on the specified port(s) until the
+        /// object is disposed. Throws if any of the ports are not free.
+        /// </summary>
+        private sealed class PortGrabber : IDisposable
```

Comment:
I'm a bit worried that this might be fragile. It's more of an integration test than a unit, and it depends on the machine that the test is running on.
If it causes problems, then it will be worth wrapping HttpListener with our own class, and mock the Start method to either fail/succeed, so we won't need to actually grab the ports on the machine.

Reply:
I considered wrapping the listener, or looking for an existing NuGet package that does it, but decided not to do it at this stage. I suspect it might become necessary as we add features.